src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -33,6 +33,44 @@ using librbd::util::create_async_context_callback;
 using librbd::util::create_context_callback;
 using librbd::util::create_rados_callback;
 
+namespace {
+
+const cls::rbd::GroupSnapshot* get_latest_mirror_group_snapshot(
+    const std::vector<cls::rbd::GroupSnapshot>& gp_snaps) {
+  for (auto it = gp_snaps.rbegin(); it != gp_snaps.rend(); ++it) {
+    const auto& snap_ns = it->snapshot_namespace;
+    auto mirror_ns = std::get_if<
+      cls::rbd::GroupSnapshotNamespaceMirror>(&snap_ns);
+    if (it->state == cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE &&
+        mirror_ns != nullptr) {
+	  return &*it;
+    }
+  }
+  return nullptr;
+}
+
+int get_group_snapshot_timestamp(librados::IoCtx& group_ioctx,
+                                 const cls::rbd::GroupSnapshot& group_snap,
+                                 utime_t* timestamp) {
+  cls::rbd::SnapshotInfo snap_info;
+  std::vector<utime_t> timestamps;
+  int r = 0;
+  for (const auto& it : group_snap.snaps) {
+    r = librbd::cls_client::snapshot_get(
```

Comment:
Future fix: this needs to use the async call.

Reply:
Noted